// angularjasper2/angular-jasper.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var AngularJasperConfig = (function () {
    function AngularJasperConfig() {
    }
    return AngularJasperConfig;
}());
exports.AngularJasperConfig = AngularJasperConfig;
var AngularJasper = (function () {
    function AngularJasper(config) {
        this.config = config;
        this.loadVisualize();
    }
    AngularJasper.prototype.loadVisualize = function () {
        var _this = this;
        var script = document.createElement('script');
        this.visualizePromise = new Promise(function (resolve, reject) {
            script.setAttribute('src', _this.config.server + "/client/visualize.js");
            script.onload = function () {
                window['visualize']({
                    auth: {
                        name: _this.config.login,
                        password: _this.config.password
                    }
                }, resolve, reject);
            };
            document.body.appendChild(script);
        });
    };
    return AngularJasper;
}());
AngularJasper = __decorate([
    core_1.Injectable(),
    __metadata("design:paramtypes", [AngularJasperConfig])
], AngularJasper);
exports.AngularJasper = AngularJasper;
