// angularjasper2/angularjasper2.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var angular_jasper_service_1 = require("./angular-jasper.service");
exports.AngularJasper = angular_jasper_service_1.AngularJasper;
exports.AngularJasperConfig = angular_jasper_service_1.AngularJasperConfig;
var report_component_1 = require("./report/report.component");
exports.ReportComponent = report_component_1.ReportComponent;
var param_directive_1 = require("./report/param.directive");
var AngularJasperModule = (function () {
    function AngularJasperModule() {
    }
    AngularJasperModule.initializeApp = function (config) {
        return {
            ngModule: AngularJasperModule,
            providers: [
                { provide: angular_jasper_service_1.AngularJasperConfig, useValue: config }
            ]
        };
    };
    return AngularJasperModule;
}());
AngularJasperModule = __decorate([
    core_1.NgModule({
        declarations: [report_component_1.ReportComponent, param_directive_1.Param],
        exports: [report_component_1.ReportComponent, param_directive_1.Param],
        providers: [angular_jasper_service_1.AngularJasper]
    }),
    __metadata("design:paramtypes", [])
], AngularJasperModule);
exports.AngularJasperModule = AngularJasperModule;
